drivers/can/can_max32.c

```diff
@@ -0,0 +1,776 @@
+/*
+ * Copyright (c) 2024-2025 Analog Devices, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_max32_can
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/can.h>
+#include <zephyr/drivers/can/transceiver.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/adi_max32_clock_control.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <can.h>
+#include <wrap_max32_can.h>
+
+LOG_MODULE_REGISTER(can_max32, CONFIG_CAN_LOG_LEVEL);
+
+#define MAX32_CAN_MAX_DLC              8
+#define MAX32_TX_CNT                   1
+#define MAX32_CAN_DRIVER_RXTX_BUF_SIZE 64
+
+/* dev_list will be used only in object_event_callback() unit_event_callback() */
+static const struct device *dev_list[MXC_CAN_INSTANCES];
+
+struct max32_req_data {
+	mxc_can_req_t req;
+	mxc_can_msg_info_t info;
+	uint8_t buf[MAX32_CAN_DRIVER_RXTX_BUF_SIZE];
+};
+
+struct max32_can_tx_callback {
+	can_tx_callback_t function;
+	void *user_data;
+};
+
+struct max32_can_rx_callback {
+	can_rx_callback_t function;
+	void *user_data;
+	struct can_filter filter;
+};
+
+struct max32_can_data {
+	struct can_driver_data common;
+
+	enum can_state state;
+	struct k_mutex inst_mutex;
+
+	struct k_sem tx_sem;
+	struct max32_can_tx_callback tx_callback;
+	struct max32_req_data tx_data;
+
+	uint8_t filter_usage;
+	struct max32_can_rx_callback rx_callbacks[CONFIG_CAN_MAX_FILTER];
+	struct max32_req_data rx_data;
+};
+
+struct max32_can_config {
+	struct can_driver_config common;
+
+	mxc_can_regs_t *can;
+	uint8_t can_id;
+
+	uint8_t irqn;
+	void (*irq_config_func)(const struct device *dev);
+
+	const struct device *clock;
+	struct max32_perclk perclk;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+static void can_max32_convert_canframe_to_req(const struct can_frame *msg, mxc_can_req_t *req)
+{
+	mxc_can_msg_info_t *info = req->msg_info;
+
+	if (msg->flags & CAN_FRAME_IDE) {
+		info->msg_id = MXC_CAN_EXTENDED_ID(msg->id);
+	} else {
+		info->msg_id = MXC_CAN_STANDARD_ID(msg->id);
+	}
+
+	info->rtr = ((msg->flags & CAN_FRAME_RTR) != 0) ? 1 : 0;
+	info->fdf = ((msg->flags & CAN_FRAME_FDF) != 0) ? 1 : 0;
+	info->brs = ((msg->flags & CAN_FRAME_BRS) != 0) ? 1 : 0;
+	info->esi = ((msg->flags & CAN_FRAME_ESI) != 0) ? 1 : 0;
```

Comment:
This is very confusing to review. Some parts of the driver seems to indicate it supporting only CAN 2.0B, while other code sections process CAN FD frame flags etc. Which is it?

Please clear this up and re-request a review.